File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n=len(nums)
        t=[0]*(n+1)
        t[1]=nums[0]
        for i in range(2,(n+1)):
            theft=nums[i-1]+t[i-2]
            skip=t[i-1]
            t[i]=max(theft,skip)
        return t[-1]
        
        

File: 0198-house-robber/0198-house-robber.java
import java.util.Arrays;
class Solution {
    public static int solve(int ind, int []arr,int []dp){
		if(ind==0){
			return arr[0];
		}
		if(ind<0){
			return 0;
		}
		if(dp[ind]!=-1){
			return dp[ind];
		}
		int pick=Solution.solve(ind-2, arr, dp)+arr[ind];
		int notPick=Solution.solve(ind-1, arr, dp);
		dp[ind]=Math.max(pick, notPick);
		return dp[ind];
	}
    
    public int rob(int[] nums) {
        int l=nums.length;
        int []dp=new int[l];
        Arrays.fill(dp,-1);
        return Solution.solve(l-1,nums,dp);
        
    }
}